specs/core/0_beacon-chain.md

```diff
@@ -1705,30 +1695,30 @@ def process_attestation(state: BeaconState, attestation: Attestation) -> None:
     attestation_slot = get_attestation_slot(state, attestation)
     assert attestation_slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot <= attestation_slot + SLOTS_PER_EPOCH
 
-    # Check target epoch, source epoch, source root, and source crosslink
     data = attestation.data
-    assert (data.target_epoch, data.source_epoch, data.source_root, data.previous_crosslink_root) in {
-        (get_current_epoch(state), state.current_justified_epoch, state.current_justified_root, hash_tree_root(state.current_crosslinks[data.shard])),
-        (get_previous_epoch(state), state.previous_justified_epoch, state.previous_justified_root, hash_tree_root(state.previous_crosslinks[data.shard])),
-    }
-
-    # Check crosslink data root
-    assert data.crosslink_data_root == ZERO_HASH  # [to be removed in phase 1]
-
-    # Check signature and bitfields
-    assert verify_indexed_attestation(state, convert_to_indexed(state, attestation))
-
-    # Cache pending attestation
     pending_attestation = PendingAttestation(
         data=data,
         aggregation_bitfield=attestation.aggregation_bitfield,
         inclusion_delay=state.slot - attestation_slot,
         proposer_index=get_beacon_proposer_index(state),
     )
+
+    assert data.target_epoch in (get_previous_epoch(state), get_current_epoch(state))
+    if data.target_epoch == get_previous_epoch(state):
+        ffg_data = (state.previous_justified_epoch, state.previous_justified_root, get_previous_epoch(state))
+        previous_crosslink = state.previous_crosslinks[data.crosslink.shard]
+        state.previous_epoch_attestations.append(pending_attestation)
     if data.target_epoch == get_current_epoch(state):
+        ffg_data = (state.current_justified_epoch, state.current_justified_root, get_current_epoch(state))
+        previous_crosslink = state.current_crosslinks[data.crosslink.shard]
         state.current_epoch_attestations.append(pending_attestation)
-    else:
-        state.previous_epoch_attestations.append(pending_attestation)
+
+    # Check FFG data, crosslink data, and signature
+    assert ffg_data == (data.source_epoch, data.source_root, data.target_epoch)
+    assert data.crosslink.epoch == min(data.target_epoch, previous_crosslink.epoch + MAX_EPOCHS_PER_CROSSLINK)
+    assert data.crosslink.previous_crosslink_root == hash_tree_root(previous_crosslink)
+    assert data.crosslink.crosslink_data_root == ZERO_HASH  # [to be removed in phase 1]
+    assert verify_indexed_attestation(state, convert_to_indexed(state, attestation))
```

Comment:
Python speaking, since `state` object is mutable, if the later assertion raises exception, the `state` (`state.previous_epoch_attestations`/`state.current_epoch_attestations`) update has been applied.

It may be a problem for pyspec when we have more complicated test cases, e.g., apply two blocks, the first one is invalid and the second one is valid. IMO putting all state updates to the last is the safest.

Reply:
I felt similarly but then looked at the rest of the doc and realized we aren't strictly doing that in all block operations

Reply:
That seems anti-pattern sin. 🙈

Reply:
@hwwhww Do you want to open an issue to discuss (and potentially fix) outside of this PR?

Reply:
@JustinDrake Yes, will do! :)

Reply:
#1059